### PR TITLE
Fix issues in classify task and add support for --data YAML format

### PR DESCRIPTION
**Pull request by Yakuho**
## fix Issue:

When training a classify model, I found that if the class labels in the validation set are not aligned with those in the training set, the validation labels become incorrect.

Because [Ultralytics:torchvision.datasets.ImageFolder](https://github.com/ultralytics/ultralytics/blob/1da4e933181f0dfe983349c2b1dcda81dba094f8/ultralytics/data/dataset.py#L726) didn't know full names when loaded validation set.

**Minimal reproducible example：**

1. download `Imagenet10` dataset.
2. change val dataset folder to:

    origin:
    
    ```
    somewhere
    └── datasets
         └── imagenet10
              └── train
              └── val # total 10 folders
                   └── n01440764
                   └── n01443537
                   └── ...
                   └── n01518878
    ```

    changed:
    
    ```
    somewhere
    └── datasets
         └── imagenet10
              └── train
              └── val # total 8 folders
                   └── n01440764
                   └── n01443537  # remove
                   └── ...
                   └── n01514668  # remove
                   └── ...
                   └── n01518878
    ```

3. debug here [Ultralytics:torchvision.datasets.ImageFolder](https://github.com/ultralytics/ultralytics/blob/1da4e933181f0dfe983349c2b1dcda81dba094f8/ultralytics/data/dataset.py#L729) and observe `torchvision.datasets.ImageFolder` attribute

    It can found that:
    
    - trainer dataset self.base.class_to_idx is:
    
    ```python
    self.base.class_to_idx == {'n01440764': 0, 'n01443537': 1, 'n01484850': 2, 'n01491361': 3, 'n01494475': 4, 'n01496331': 5, 'n01498041': 6, 'n01514668': 7, 'n01514859': 8, 'n01518878': 9}
    ```
    
    - val dataset self.base.class_to_idx is:
    
    ```python
    self.base.class_to_idx == {'n01440764': 0, 'n01484850': 1, 'n01491361': 2, 'n01494475': 3, 'n01496331': 4, 'n01498041': 5, 'n01514859': 6, 'n01518878': 7}
    ```

### Solution:

I noticed that [This code snippet](https://github.com/ultralytics/ultralytics/blob/da1080ecaae1cd8a0be17df9d9a70a8a6d80e4ff/ultralytics/models/yolo/classify/train.py#L141) have been solved some compatibility issues, but it does not work for the case mentioned above.

Therefore, I resolved this issue by passing `data["names"]` from [def check_cls_dataset -> result](https://github.com/Yakuho/ultralytics/blob/d88598e24a77d7367e6659d973659c0a82040ffc/ultralytics/data/utils.py#L480)
to [ClassificationDataset](https://github.com/Yakuho/ultralytics/blob/d88598e24a77d7367e6659d973659c0a82040ffc/ultralytics/data/dataset.py#L712) and remapping the labels accordingly [here](https://github.com/Yakuho/ultralytics/blob/d88598e24a77d7367e6659d973659c0a82040ffc/ultralytics/data/dataset.py#L738).

## New Feature:

In addition, I also added support for the YAML format to train classify task.

**Minimal reproducible example for `imagenet10`：**

```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) list: [path/to/imgs1, path/to/imgs2, ..]
path: imagenet10 # dataset root dir or dataset name
train: train # train images (relative to 'path') 12 images
val: val # val images (relative to 'path') 12 images
test: # test images (optional)

# Classes
names:
  0: tench
  1: goldfish
  2: great white shark
  3: tiger shark
  4: hammerhead shark
  5: electric ray
  6: stingray
  7: cock
  8: hen
  9: ostrich
```

1. `Path` has been set but not a path dir (depends contains '/' or not e.g. `"imagenet"`)
  behaves: `Ultralytics` will download dataset from [download(ultralytics/utils/downloads.py)](https://github.com/ultralytics/ultralytics/blob/da1080ecaae1cd8a0be17df9d9a70a8a6d80e4ff/ultralytics/utils/downloads.py#L481) by keyword (`path`), and `train_set`, `val_set`, and `test_set` are appended to `path`
2. `Path` has been set and is a path dir (contains '/' e.g., `/nasdata/path/to/dataset`)
  behaves: `train_set`, `val_set`, and `test_set` are appended to `path`, e.g. `/nasdata/path/to/dataset/train`
3. `Path` has been set null
  behaves: `train_set`, `val_set`, and `test_set` are treated as full paths when accessed, so those `data_set` should set e.g. `train_set: \path\to\dataset\train`
4. `train_set`, `val_set`, and `test_set` are support set more than one path and each subpath dataset format still as [origin format](https://docs.ultralytics.com/zh/datasets/classify/#folder-structure-example) style. It is noting that label class will get union if set more than one path. example:
    ```
    somewhere
    └── datasets
         └── example_dataset
              └── train0
                   └── class-00
                   └── class-01
                   └── class-02
                   └── class-03
              └── train1
                   └── class-00
                   └── class-04
                   └── class-05
                   └── class-06
    ```
    - If `names` was set in the YAML (as a list or dict), model class names are loaded in that order:
      ```yaml
      path: example_dataset
      train:  
        - train0 
        - train1
      val: val 
      test: # test images (optional)
       
      names: ["class-00", "class-01", "class-02", "class-03", "class-04", "class-05", "class-06"]  # as list
      names: {0: 'class-00', 1: 'class-01', 2: 'class-02', 3: 'class-03', 4: 'class-04', 5: 'class-05', 6: 'class-06'}  # as dict
      names: # or set null
      ```
    - otherwise, they are sorted automatically based on subfolder names:
      ```python
      sorted(list(set(n.name for sf in map(Path, ("/path/to/train0", "/path/to/train1")) if sf.is_dir() for n in sf.glob("*") if n.is_dir())))
      ```
5. `names` support `dict` and `list`, as explained above. It can also be set to null; when `names` is empty, all labels under the `train_set` directories are automatically collected as a union and sorted to `names`. Beside, model train task can select only the labels to include for training via the YAML configuration. example:
    ```yaml
    names: ["class-00", "class-01", "class-02", "class-03", "class-04"]  # as list
    names: {0: 'class-00', 1: 'class-01', 2: 'class-02', 3: 'class-03', 4: 'class-04'}  # as dict
    # the other label [class-05, class-06] will be ignore
    ```

## Declaration:

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves classification dataset handling and adds YAML-based `--data` support for classify workflows 🧩📄

### 📊 Key Changes
-Adds `ImageNet10.yaml` dataset config and updates `ImageNet.yaml` guidance for directory/list-based splits.
-Extends `check_cls_dataset()` to accept `--data` as a YAML file, resolving paths/splits (including list-of-roots) and supporting dataset-name downloads.
-Refactors `ClassificationDataset` to accept `names`, support multiple roots, align `ImageFolder` class indices to provided names, and update caching/verification for multi-subdataset scans.
-Fixes validator routing so classification uses `check_cls_dataset()` even when `--data` is a YAML file.
-Removes train-time filtering workaround in `classify/train.py` now that dataset class alignment is handled centrally.

### 🎯 Purpose & Impact
-Enables `yolo train task=classify data=...yaml` workflows and consistent dataset configuration across tasks ✅
-Prevents CUDA assertion errors caused by mismatched class indices by enforcing alignment to `names` at dataset construction time 🔒
-Adds a tiny ImageNet10 config for fast smoke tests/debugging and improves dataset download/scan behavior for multi-root splits ⚡

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Yakuho, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review—an Ultralytics engineer will assist shortly. Please review the following checklist:

-✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

PR looks clean overall. I didn’t find any definite bugs, performance regressions, or missing error handling in the diff. Nice improvements to classification dataset handling and YAML-based `data` support.
